ui/org.eclipse.pde.core/src/org/eclipse/pde/internal/core/SourceLocationManager.java

```diff
@@ -446,6 +463,27 @@ protected BundleManifestSourceLocationManager initializeBundleManifestLocations(
 
 	private static final class SourceExtensions {
 		final Collection<SourceLocation> locations = new LinkedHashSet<>();
-		final List<IPluginSourcePathLocator> locators = new ArrayList<>();
+		final List<OrderedPluginSourcePathLocator> locators = new ArrayList<>();
```

Comment:
Can't you just make this a `List<Map.Entry<IPluginSourcePathLocator, LocatorComplexity>>` and pass a comparator to `Collections.sort()` like `Comparator.comparing(Entry::getValue)`?
That would save the extra inner static class.

Reply:
I think we do not safe much here and are prepared to further enhancements e.g. I think the next would be to have some kind of priority to sort items on the same complexity level, e.g. as you said naming conventions might be only a weak assumption as you said and we probably want to prefer other metrics first, also this eases debugging.